### ISSUE TITLE
Define what the terms "shutdown", "terminate" and "kill" mean

### ISSUE THREAD

**Issue by e2** [celluloid/celluloid#557]
#### TL;DR - I know "shutting things down" is a neglected area, but this has to change for sanity during debugging, development and testing.

The idea is so the definition matches the usage - and the usage matches the definition.

E.g. Tasks should be killed only as an absolute last result - e.g. only when they are incapable of responding to a Termination request.

Pretty much the same with `join` - it's important to define whether shutdown waits or not (e.g. terminating an actor just sets `running=false`), and whether it joins to actors or not.

Either way these terms should be applied consistently, whether it's an actor, a task pool, an actor pool, a supervision group, the actor system and even Celluloid itself.

This also means defining which can be interrupted etc.

Some examples:
- every kill should be well warranted (only as a last resort), unless it's e.g. an idle thread waiting for tasks
- every join should have a timeout (to avoid deadlocks)
- even during shutdown - actors should be allowed to create new tasks to properly clean up after themselves
- there needs to be a distinction between a "normal shutdown" and otherwise (like `exit` vs `abort`)
- and there error/exception handling as well

Currently, a "clean shutdown" is not possible in Celluloid.

**Comment by tarcieri** [comment]
`#terminate` is analogous to Akka's `PoisonPill`: it sends a message to an actor requesting it to terminate as soon as the message is process.

`#kill` is an unsafe API ala `Thread.kill` that should generally not be used.

**Comment by e2** [comment]
> kill is an unsafe API ala Thread.kill that should generally not be used.

100% agree. In fact, I'd only use it in a signal handler - if at all.

Let's just take actor_system as an example.

Here's a list of areas I'm not comfortable with:

1.`actors = running`, which takes a snapshot
2. no mutex, so shutdown could be called multiple times
3. creates a single all-encompassing Timeout
4. uses actors.size, which is unprotected by a mutex
5. first calls `terminate!`, which is the proper way to terminate an actor (TerminationRequest on mailbox)
6. rescues dead actor error without removing actor from the queue, though it's strange to get a ref to an actor with a dead mailbox in the first place
7. then it potentially deadlocks temporarily by calling Actor.join (with no timeout)
8. if one such actor deadlocks, no other receives a join - and also is unable to cleanly report a list of exceptions before accepting the TerminationRequest
9. Also rescuing DeadActor - which probably doesn't make sense
10. Timeout::Error prevents permanent deadlocks, but then indiscriminately kills the threads and kills their mailboxes (also silently rescuing DeadActorError and MailboxDead)
11. At this point, I'm unsure about actor finalizers (may be ok there)
12. the ensure block calls `(group/pool).shutdown`, which is pretty much the same above scenario, except for threads, not actors
13. Individually, the shutdown doesn't have a timeout parameter - though it's probably not worth implementing
14. there shouldn't be any errors or warnings during a clean shutdown
15. obviously, the first to be killed should be supervises which respawn

To me - if actors are unable to cleanly free their resources (whatever they may be), it's not a clean shutdown.

My question though: is shutdown supposed to be clean? or is it just a async "request to terminate"?

I think I just need a few terms:
1. "signal everything to finish and wait - or fail" (shutdown)
2. "signal everything to finish - async + wait on at_exit if necessary" (terminate)
3. "do your very best to shut down within a given amount of time (or fail if timeout cannot be fulfilled)" (shutdown with timeout)
4. same as 3, but resort to killing if necessary (shutdown! or terminate!)
5. "kill immediately due to fatal error" (kill)

Do those make sense?

Further more, I think shutdown should potentially throw a variety of errors - it would be up to the user to decide what to do with them (log, ignore, retry, etc.).

**Comment by tarcieri** [comment]
@e2 all of that is really reaping logic.

A true clean shutdown should begin by terminating the supervision hierarchy. Unfortunately, it seems like the supervision hierarchy termination code has likewise been lost along the way. Le sigh.

**Comment by e2** [comment]
@tarcieri - yes, exactly - I think the best way to go about this is replacing kill() and unchecked joins() with better reaping logic (it's a start at least). 

Just avoiding kill() and mindless joins() will help encounter and fix issues like this one: https://github.com/celluloid/celluloid/issues/333

I'll see if I can extract/mock the shutdown sequences so that they can properly be unit tested.

I think the task groups (Spawner/Pool) need a "friendly shutdown" before anything else happens.

The general order of shutting things down would be:
1. supervisors of supervisors
2. supervisors
3. pools
4. single/user actors
6. celluloid-core actors
7. leftover actor threads
6. leftover user task threads/fibers (can be killed earlier)
7. leftover task threads/fibers

Did I miss anything?

**Comment by tarcieri** [comment]
In the before time, in the long long ago, before getting lost in the sands of time, calling `Celluloid::SupervisionGroup.run` would spawn the root of the supervision tree, and set this accessor to be the root of the supervision tree:

https://github.com/celluloid/celluloid/blob/master/lib/celluloid/supervisor.rb#L7

Unfortunately grep confirms all of that code has been lost.

Shutdown should start by checking if the supervision tree has a root, and if it does, terminating the root supervisor.

SupervisionGroups walk their children, shutting them down in the reverse order they started in, and doing this synchronously:

https://github.com/celluloid/celluloid/blob/master/lib/celluloid/supervision_group.rb#L177

**Comment by digitalextremist** [comment]
@tarcieri is this the behavior we want today-now, in #511?

**Comment by tarcieri** [comment]
No, I just added it: "Add back support for supervision trees"

**Comment by e2** [comment]
I'm closing this, because it's covered by other issues/work, such as:

- removing `kill` method on JRuby + Rubinius
- removing Actor#join calls in specs (done - but the API needs more thinking)
- Reaper (not ready - and won't be production ready for 0.17.0)
- hardening the specs (dealing with "dangerously-suspending" warning before continuing with shutdown logic)

**Comment by openface** [comment]
I need the ability to do a "friendly termination", where sending a QUIT signal to my application causes all of the running actors to wait and shutdown.   What issue should I be participating/watching for that feature?   Thanks.

**Comment by digitalextremist** [comment]
We'll find the code that allowed clean shutdown ( it was refactored out somewhere ) and get it back in by `0.17.5` ( we're pushing the release before that right now ) -- shouldn't be too long.

**Comment by tarcieri** [comment]
I feel like Java's ThreadPoolExecutors are a good place to look for this. Specifically they support the sort of mechanisms @e2 has been describing: a way to drain the pool of work and wait (with a prescribed timeout) for that to happen (or timeout)

**Comment by e2** [comment]
@openface - you can simply set `Celluloid.shutdown_timeout=1_000_000`, and call `Celluloid.shutdown` in your `QUIT` signal trap. 

Of course this will prevent your app from exiting if an actor is stuck waiting forever on something - so it's up to you to put custom timeouts in such places.

There are ways to "try to detect" if an actor is sleeping or waiting - but this gets complex fast. It's best to have an actor that can signal a busy actor to finish (while the busy actor should be designed to respond to such a signal).

**Comment by openface** [comment]
Thanks @e2, that's precisely what I'm hoping to do- however I'm not able to make this work, and I'm not sure if the problem is in Celluloid, or in my code.

My unicorn worker has this signal handler:
````
# Wait up to 5 minutes for jobs to finish before exiting
Celluloid.shutdown_timeout = 300
at_exit do
  Celluloid.shutdown
end
````

To test this behavior, I have an actor sleeping for 60 seconds.   When I send a QUIT signal to the application, I get:

````
Terminating 3 actors...
Terminating task: type=:call, meta={:method_name=>:perform}, status=:callwait
	Celluloid::TaskFiber backtrace unavailable. Please try `Celluloid.task_class = Celluloid::TaskThread` if you need backtraces here.
Celluloid::PoolManager: async call `perform` aborted!
Celluloid::Task::TerminatedError: task was terminated
	/u/jhines/.rvm/gems/ruby-2.1.1@burt_api_server/gems/celluloid-0.16.0/lib/celluloid/tasks/task_fiber.rb:34:in `terminate'
	/u/jhines/.rvm/gems/ruby-2.1.1@burt_api_server/gems/celluloid-0.16.0/lib/celluloid/actor.rb:345:in `each'
	/u/jhines/.rvm/gems/ruby-2.1.1@burt_api_server/gems/celluloid-0.16.0/lib/celluloid/actor.rb:345:in `cleanup'
	/u/jhines/.rvm/gems/ruby-2.1.1@burt_api_server/gems/celluloid-0.16.0/lib/celluloid/actor.rb:329:in `shutdown'
	/u/jhines/.rvm/gems/ruby-2.1.1@burt_api_server/gems/celluloid-0.16.0/lib/celluloid/actor.rb:164:in `run'
	/u/jhines/.rvm/gems/ruby-2.1.1@burt_api_server/gems/celluloid-0.16.0/lib/celluloid/actor.rb:130:in `block in start'
	/u/jhines/.rvm/gems/ruby-2.1.1@burt_api_server/gems/celluloid-0.16.0/lib/celluloid/thread_handle.rb:13:in `block in initialize'
	/u/jhines/.rvm/gems/ruby-2.1.1@burt_api_server/gems/celluloid-0.16.0/lib/celluloid/actor_system.rb:32:in `block in get_thread'
	/u/jhines/.rvm/gems/ruby-2.1.1@burt_api_server/gems/celluloid-0.16.0/lib/celluloid/internal_pool.rb:130:in `call'
	/u/jhines/.rvm/gems/ruby-2.1.1@burt_api_server/gems/celluloid-0.16.0/lib/celluloid/internal_pool.rb:130:in `block in create'
Terminating task: type=:finalizer, meta={:method_name=>:__shutdown__}, status=:receiving
	Celluloid::TaskFiber backtrace unavailable. Please try `Celluloid.task_class = Celluloid::TaskThread` if you need backtraces here.
Terminating task: type=:call, meta={:method_name=>:perform}, status=:sleeping
	Celluloid::TaskFiber backtrace unavailable. Please try `Celluloid.task_class = Celluloid::TaskThread` if you need backtraces here.
````

The application immediately exits, and the actor never completes it's work.  It doesn't seem to be affected by the extended timeout period.

**Comment by e2** [comment]
@openface - if you really want an actor to sleep (for testing), make sure you're using `Kernel.sleep` to avoid using Celluloid's overriding `sleep` method. Or, you can use a busy loop for testing. 

If you're still having problems, you can setup a repo with the problem.

Details (if you're interested):

In your example calling `sleep` just suspended the actor (preventing it from being blocked) and `Actor.terminate!` was called (by the `shutdown` method), which sets the actor's running state to `false`. 

Shortly afterwards, the actor gets the termination request and cleans up after itself - by murdering any running tasks (like your `process` method), and that's why you get this warning:

```
Terminating task: type=:call, meta={:method_name=>:perform}, status=:callwait
```

NOTE: Using `Kernel.sleep` is a "hack" to block the actor from receiving new messages (like the termination request).

**Comment by openface** [comment]
Thanks again @e2.  I can see the actors do complete when I call `exec('sleep 30')`, but not with `sleep` or `Kernel.sleep`.  

However, I think I am not clearly explaining my intent- which is more related to how to tell Celluloid to wait for unprocessed/queued actors in the pool.  It looks like a call to `Celluloid.shutdown` will tell any not-yet-run actors to never run.

**Comment by e2** [comment]
@openface - I think the only way is to hack/monkey-patch Celluloid so it doesn't call shutdown (or delay it until your actors are finished). 

I'd say this is currently the *recommended* solution at this moment. (The current shutdown sequence is pretty "brutal" - and that won't change soon). The shutdown logic isn't that complex really - so don't feel intimidated.

What you want to do is wait until the important actors are finished, and *then* call shutdown. 

It's up to you do define what "finished" means, wait for that "finished" to happen, and then you can either call `Celluloid.shutdown` or shutdown actors manually (and then shutdown the thread pool once no more actors are running).

I think it's best if you prepare a small repo with the problem - since the "best" solution may strongly depend on what you want to accomplish. Ideally, open a new issue (and link your example repo) - so it can be turned into a test case for the future.
